collector/lambdacomponents/custom.go

```diff
@@ -0,0 +1,127 @@
+//go:build lambdacomponents.custom
+
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lambdacomponents
+
+import (
+	_ "github.com/open-telemetry/opentelemetry-lambda/collector/lambdacomponents/connector"
+	_ "github.com/open-telemetry/opentelemetry-lambda/collector/lambdacomponents/exporter"
+	_ "github.com/open-telemetry/opentelemetry-lambda/collector/lambdacomponents/extension"
+	_ "github.com/open-telemetry/opentelemetry-lambda/collector/lambdacomponents/processor"
+	_ "github.com/open-telemetry/opentelemetry-lambda/collector/lambdacomponents/receiver"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/connector"
+	"go.opentelemetry.io/collector/exporter"
+	"go.opentelemetry.io/collector/extension"
+	"go.opentelemetry.io/collector/otelcol"
+	"go.opentelemetry.io/collector/processor"
+	"go.opentelemetry.io/collector/receiver"
+
+	"go.uber.org/multierr"
+)
+
+type factory[T any] interface {
+	Get(extensionId string) T
+}
+
+type FactoryFn[T any] func() T
+
+func (fn FactoryFn[T]) Get(extensionId string) T {
+	return fn()
+}
+
+type FactoryWithExtensionIdFn[T any] func(extensionId string) T
+
+func (fn FactoryWithExtensionIdFn[T]) Get(extensionId string) T {
+	return fn(extensionId)
+}
+
+var (
+	receiverFactories  []factory[receiver.Factory]
+	processorFactories []factory[processor.Factory]
+	exporterFactories  []factory[exporter.Factory]
+	extensionFactories []factory[extension.Factory]
+	connectorFactories []factory[connector.Factory]
```

Comment:
can we simplify this? can this be a list of functions (providers) that receive a extensionid as parameter and then return a factory as return value?

```
	receiverFactories  []func(extensionid string) receiver.Factory
	processorFactories []func(extensionid string) processor.Factory
	exporterFactories  []func(extensionid string)  exporter.Factory
	extensionFactories []func(extensionid string)  extension.Factory
	connectorFactories []func(extensionid string) connector.Factory
```

Then in each init block in the modules you can use the `AddReceiverFactoryProvider` & etc to append to the slices.

e.g.:
```
func init() {
	lambdacomponents.AddReceiverFactoryProvider(func(extensionId string) receiver.Factory { return telemetryapireceiver.NewFactory(extensionID) })
}
```
then in `Components` you just iterate over the elements invoking the provider to get the factories and pass that to the appropriate `MakeFactoryMap`.

It think what you did is very nice but I prefer to keep the code simpler so that it is simpler to maintain and more welcoming for people trying to understand the code to extend the collector layer using your approach.